Translate ch17-02-trait-objects.md via GitLocalize (#1133)
rust-lang-ru/book: rustbook-ru/src/ch17-02-trait-objects.md

Однако иногда мы хотим, чтобы пользователь нашей библиотеки мог расширить набор типов, которые допустимы в конкретной ситуации. Чтобы показать как этого добиться, мы создадим пример инструмента с графическим интерфейсом пользователя (GUI), который просматривает список элементов, вызывает метод `draw` для каждого из них, чтобы нарисовать его на экране - это обычная техника для инструментов GUI. Мы создадим библиотечный крейт с именем `gui`, содержащий структуру библиотеки GUI. Этот крейт мог бы включать некоторые готовые типы для использования, такие как `Button` или `TextField`. Кроме того, пользователи такого крейта `gui` захотят создавать свои собственные типы, которые могут быть нарисованы: например, кто-то мог бы добавить тип `Image`, а кто-то другой добавить тип `SelectBox`.

> В этом примере мы не будем реализовывать полноценную библиотеку GUI, но покажем, как некоторые её части будут сочетаться друг с другом. На момент написания библиотеки мы не имеем представления о всех тех типах, которые могут захотеть создать другие программисты и соответственно не можем их определить. Это означает, что `gui` должен поддерживать множество различных типов и мочь вызывать метод `draw` у каждого из них. Ему не нужно точно знать, что произойдет, когда мы вызовем метод `draw`, достаточно знать, что у того или иного объекта будет доступен этот метод для вызова.

Reviewer: ```suggestion
Мы не будем реализовывать полноценную библиотеку GUI для этого примера, но покажем, как её части будут подходить друг к другу. На момент написания библиотеки мы не можем знать и определить все типы, которые могут захотеть создать другие программисты. Но мы знаем, что `gui` должен отслеживать множество значений различных типов и ему нужно вызывать метод `draw` для каждого из этих значений различного типа. Ему не нужно точно знать, что произойдёт, когда вызывается метод `draw`, просто у значения будет доступен такой метод для вызова.
```

Чтобы сделать это на языке с наследованием, можно определить класс с именем `Component` у которого есть метод с названием `draw`. Другие классы, такие как `Button`, `Image` и `SelectBox` наследуются от `Component` и следовательно, наследуют метод `draw`. Каждый из них может переопределить реализацию метода `draw`, чтобы определить своё пользовательское поведение, но платформа может обрабатывать все типы, как если бы они были экземплярами `Component` и вызывать `draw` у них. Но поскольку в Rust нет наследования, нам нужен другой способ структурировать `gui` библиотеку, чтобы позволить пользователям расширять её новыми типами.

### Определение типажа для общего поведения

> Чтобы реализовать поведение, которое мы хотим видеть у `gui`, мы определим трейт `Draw`, который будет иметь один метод `draw`. Затем мы определим вектор, который принимает объект *trait*. Объект трейта указывает как на экземпляр типа, реализующего указанный нами трейт, так и на таблицу, используемую для поиска методов трейта на этом типе во время выполнения. Мы создаем трейт-объект, используя какой-либо указатель, например, ссылку `&` или умный указатель `Box<T>`, затем ключевое слово `dyn`, а затем указываем соответствующий трейт. (О том, почему для трейт-объектов необходимо использовать указатель, мы поговорим в главе 19 в разделе ["Динамически изменяемые типы и трейт `Sized`]). Мы можем использовать трейт-объекты в качестве замены универсального или конкретного типа. Где бы мы ни использовали трейт-объект, система типов Rust гарантирует во время компиляции, что любое значение, используемое в этом контексте, будет реализовывать трейт этого трейт-объекта. Следовательно, нам не нужно знать все возможные типы во время компиляции.

Reviewer: ```suggestion
Чтобы реализовать поведение, которое мы хотим иметь в `gui`, определим типаж с именем `Draw`, который будет содержать один метод с названием `draw`. Затем мы можем определить вектор, который принимает *типаж-объект*. Типаж-объект указывает как на экземпляр типа, реализующего указанный типаж, так и на внутреннюю таблицу, используемую для поиска методов типажа указанного типа во время выполнения. Мы создаём типаж-объект в таком порядке: используем какой-нибудь вид указателя, например ссылку `&` или умный указатель `Box<T>`, затем ключевое слово `dyn`, а затем указываем соответствующий типаж. (Мы будем говорить о причине того, что типаж-объекты должны использовать указатель в разделе ["Типы динамического размера и типаж `Sized` "] &lt;!-- <!-- --> --&gt; главы 19). Мы можем использовать типаж-объекты вместо универсального или конкретного типа. Везде, где мы используем типаж-объект, система типов Rust проверит во время компиляции, что любое значение, используемое в этом контексте, будет реализовывать нужный типаж у типаж-объекта. Следовательно, нам не нужно знать все возможные типы во время компиляции.
```

<span class="caption">Листинг 17-3: Определение типажа <code>Draw</code></span>

> Этот синтаксис вам должен быть знаком по прошлому обсуждению определения трейтов в главе 10. Теперь перейдем к новому синтаксису: В листинге 17-4 определена структура с именем `Screen`, которая содержит вектор с названием `components`. Этот вектор представлен типом `Box<dyn Draw>`, который является трейт-объектом - он представляет собой обозначение любого типа внутри `Box`, реализующего трейт `Draw`.

Reviewer: ```suggestion
Этот синтаксис должен выглядеть знакомым из наших дискуссий о том, как определять типажи в главе 10. Далее следует новый синтаксис: в листинге 17.4 определена структура с именем `Screen`, которая содержит вектор с именем `components`. Этот вектор имеет тип `Box<dyn Draw>`, который и является типаж-объектом; это замена для любого типа внутри `Box` который реализует типаж `Draw`.
```

<span class="caption">Листинг 17-9: Использование типаж-объектов для хранения значений разных типов, реализующих один и тот же типаж</span>

> Когда мы писали библиотеку, мы не предполагали, что кто-то добавит тип `SelectBox`, но наша реализация `Screen` смогла работать с новым типом и отрисовывать его, потому что `SelectBox` реализует признак `Draw`, что означает, что он реализует метод `draw`.

Reviewer: ```suggestion
Когда мы писали библиотеку, мы не знали, что кто-то может добавить тип `SelectBox`, но наша реализация `Screen` могла работать с новым типом и рисовать его, потому что `SelectBox` реализует типаж `Draw`, что означает, что он реализует метод `draw`.
```

> Эта концепция - опираться не на конкретный тип, а только на определённые факторы которым объект соответствует - похожа на принцип *duck typing* в динамически типизированных языках: если нечто ходит как утка и крякает как утка, то это, вероятно, и есть утка! В реализации `run` у `Screen` в листинге 17-5, `run` не нужно знать конкретный тип каждого компонента. Он не проверяет, является ли компонент экземпляром `Button` или `SelectBox`, он просто вызывает метод `draw` для компонента. Указав `Box<dyn Draw>` в качестве типа элементов вектора `components`, мы определили `Screen` как объекты, у которых можно вызвать метод `draw`.

Reviewer: ```suggestion
Эта концепция, касающаяся только сообщений на которые значение отвечает, в отличии от конкретного тип у значения, аналогична концепции *duck typing* в динамически типизированных языках: если что-то ходит как утка и крякает как утка, то она должна быть утка! В реализации метода `run` у `Screen` в листинге 17-5, `run` не нужно знать каким будет конкретный тип каждого компонента. Он не проверяет, является ли компонент экземпляром `Button` или `SelectBox`, он просто вызывает метод `draw` компонента. Указав `Box<dyn Draw>` в качестве типа значений в векторе `components`, мы определили `Screen` для значений у которых мы можем вызвать метод `draw`.
```

Преимущество использования типаж-объектов и системы типов Rust для написания кода, похожего на код с использованием концепции duck typing состоит в том, что нам не нужно во время выполнения  проверять реализует ли значение в векторе конкретный метод или беспокоиться о получении ошибок, если значение не реализует метод, мы все равно вызываем метод. Rust не скомпилирует наш код, если значения не реализуют типаж, который нужен типаж-объектам.

> Например, в листинге 17-10 показано, что произойдет, если мы попытаемся создать `Screen` с `String` в качестве его компонента:

Reviewer: ```suggestion
Например, в листинге 17-10 показано, что произойдёт, если мы попытаемся создать `Screen` с `String` в качестве его компонента:
```

### Типаж-объекты выполняют динамическую диспетчеризацию (связывание)

> Вспомните, в разделе "Производительность кода, использующего универсальные типы" в главе 10 наше обсуждение процесса мономорфизации, выполняемого компилятором, когда мы используем ограничения трейтов для универсальных типов: компилятор генерирует частные реализации функций и методов для каждого конкретного типа, который мы применяем для параметра универсального типа. Код, который получается в результате мономорфизации, выполняет статическую диспетчеризацию, то есть когда компилятор знает, какой метод вы вызываете во время компиляции. Это противоположно динамической диспетчеризации, когда компилятор не может определить во время компиляции, какой метод вы вызываете. В случае динамической диспетчеризации компилятор формирует код, который во время выполнения определит, какой метод нужно вызвать.

Reviewer: ```suggestion
Напомним, в разделе [«Производительность кода с использованием обобщений»](ch10-01-syntax.html#performance-of-code-using-generics)<!-- --> главы 10 обсуждается процесс мономорфизации выполняемый компилятором, когда мы используем ограничения типажей для обобщённых типов: компилятор генерирует конкретные реализации функций и методов для каждого конкретного типа, который мы используем вместо параметра обобщённого типа. Код, полученный в результате мономорфизации, выполняет *статическую диспетчеризацию*, когда компилятор знает какой метод вы вызываете во время компиляции. Это противоположно подходу *динамической диспетчеризации*, когда компилятор не может сказать во время компиляции, какой метод вы вызываете. В случаях динамической диспетчеризации компилятор генерирует код, который во время выполнения определяет, какой метод необходимо вызывать.
```

Когда мы используем типаж-объекты, Rust должен использовать динамическую диспетчеризацию. Компилятор не знает всех типов, которые могут быть использованы с кодом, использующим типаж-объекты, поэтому он не знает, какой метод реализован для какого типа при вызове. Вместо этого, во время выполнения, Rust использует указатели внутри типаж-объекта, чтобы узнать какой метод вызвать. Такой поиск вызывает дополнительные затраты во время исполнения, которые не требуются при статической диспетчеризации. Динамическая диспетчеризация также не позволяет компилятору выбрать встраивание кода метода, что в свою очередь делает невозможными некоторые оптимизации. Однако мы получили дополнительную гибкость в коде, который мы написали в листинге 17-5, и которую смогли поддержать в листинге 17-9, поэтому все "за" и "против" нужно рассматривать в комплексе.

> ["Динамически изменяемые типы и трейт `Sized`]: ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait

Reviewer: ```suggestion
["Типы динамического размера и типаж `Sized` "]: ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait
```
